internal/fs/inode/type_cache.go

```diff
@@ -50,58 +87,126 @@ type typeCache struct {
-////////////////////////////////////////////////////////////////////////
-
-// Panic if any internal invariants have been violated. The careful user can
-// arrange to call this at crucial moments.
-func (tc *typeCache) CheckInvariants() {
-	tc.entries.CheckInvariants()
+	return &typeCache{}
 }
 
 // Insert inserts a record to the cache.
 func (tc *typeCache) Insert(now time.Time, name string, it Type) {
-	// Are we disabled?
-	if tc.ttl == 0 {
-		return
+	if tc.entries != nil {
+		_, err := tc.entries.Insert(name, cacheEntry{
+			expiry:    now.Add(tc.ttl),
+			inodeType: it,
+		})
+		if err != nil {
+			panic(fmt.Errorf("failed to insert entry in typeCache: %v", err))
+		}
+		logger.Debugf("TypeCache: Inserted %s as %s", name, it.String())
 	}
-	tc.entries.Insert(name, cacheEntry{
-		expiry:    now.Add(tc.ttl),
-		inodeType: it,
-	})
 }
 
 // Erase erases all information about the supplied name.
 func (tc *typeCache) Erase(name string) {
-	tc.entries.Erase(name)
+	if tc.entries != nil {
+		tc.entries.Erase(name)
+		logger.Debugf("TypeCache: Erased entry for %s", name)
+	}
 }
 
 // Get gets the record for the given name.
 func (tc *typeCache) Get(now time.Time, name string) Type {
+	if tc.entries == nil {
+		return UnknownType
+	}
+
+	logger.Debugf("TypeCache: Fetching entry for %s ...", name)
+
 	val := tc.entries.LookUp(name)
 	if val == nil {
+		logger.Debugf("                                     ... Not found!")
 		return UnknownType
 	}
 
 	entry := val.(cacheEntry)
 
+	logger.Debugf("                                     ... Found as %s", entry.inodeType.String())
+
 	// Has the entry expired?
 	if entry.expiry.Before(now) {
+		logger.Tracef("TypeCache: Erasing entry for %s because of TTL expiration", name)
 		tc.entries.Erase(name)
 		return UnknownType
 	}
 	return entry.inodeType
 }
+
+// A cache that wraps over a TypeCache and
+// prepends every cache entry's name/key with the supplied bucketName
+// for every operation (Insert/Get/Erase).
+//
+// Must be created with NewTypeCacheBucketView only. May be contained in a larger struct.
+// External synchronization is required.
+type typeCacheBucketView struct {
+	sharedTypeCache TypeCache
+	bucketName      string
+}
+
+func (tcbv *typeCacheBucketView) key(name string) string {
+	if tcbv.bucketName != "" {
+		return tcbv.bucketName + "/" + name
+	}
+	return name
+}
+
+////////////////////////////////////////////////////////////////////////
+// Public interface
+////////////////////////////////////////////////////////////////////////
+
+// Creates a new typeCacheBucketView wrapping over the
+// given TypeCache and prefix bucketName for prepending
+// in operational keys.
+// This is needed in case of multi-bucket mount (i.e. dynamic-mount).
+func NewTypeCacheBucketView(stc TypeCache, bn string) TypeCache {
+	if stc == nil {
+		panic("The passed shared-type-cache is nil")
+	}
+	return &typeCacheBucketView{sharedTypeCache: stc, bucketName: bn}
+}
+
+// Insert inserts a record to the cache.
```

Comment:
nit: remove these comments as we have already added them above and the name is implicit